### PR TITLE
Feature/testcases med po bills

### PR DESCRIPTION
**Pull request by tariq-9484**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -64,8 +65,6 @@ end
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
-gem 'breadcrumbs_on_rails', '4.0.0'
```
Comment (sulemanuzair) [inline Gemfile]: revert

File: Gemfile [hunk 2]
```diff
@@ -38,7 +39,7 @@ gem 'breadcrumbs_on_rails', '=4.0.0'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
-  gem 'pry','=0.13.1'
+  gem 'pry','~> 0.13.1'
```
Comment (sulemanuzair) [inline Gemfile]: revert

File: app/controllers/bills_controller.rb [hunk 3]
```diff
@@ -1,6 +1,6 @@
 class BillsController < ApplicationController
   load_and_authorize_resource find_by: :sequence_num, through: :current_hospital
-  before_action :root_page_breadcrumb, only: [:index, :new, :show, :edit]
```
Comment (sulemanuzair) [inline app/controllers/bills_controller.rb:3]: why removing this ?

File: app/controllers/bills_controller.rb [hunk 4]
```diff
@@ -36,8 +35,7 @@ def get_medicine
   def add_medicine
     @medicine= current_hospital.medicines.find_by(id: params[:medicine_id])
     quantity=params[:quantity].to_i
-    if @bill.add_medicine(@medicine,quantity) 
-      flash[:notice] = t('sales_order.addmed.success')  
+    if @bill.add_medicine(@medicine,quantity)
```
Comment (sulemanuzair) [inline app/controllers/bills_controller.rb]: space (@medicine, quantity)

File: app/controllers/purchase_order_controller.rb [hunk 5]
```diff
@@ -23,7 +23,6 @@ def get_medicine
     if @medicine.blank?
       flash[:notice] = t('medicine.search.failure')
     else
-      flash[:notice] = t('medicine.search.success')
       respond_to do |format|
           format.js{ render 'search'  }
```
Comment (sulemanuzair) [inline app/controllers/purchase_order_controller.rb]: space after format.js
format.js { render 'search' }

File: app/controllers/purchase_order_controller.rb [hunk 6]
```diff
@@ -33,8 +32,7 @@ def get_medicine
   def add_medicine
     @medicine= current_hospital.medicines.find_by(id: params[:medicine_id])
     quantity=params[:quantity].to_i
-    if @purchase_order.add_medicine(@medicine,quantity) 
-      flash[:notice] = t('purchase_order.addmed.success')  
+    if @purchase_order.add_medicine(@medicine,quantity)
```
Comment (sulemanuzair) [inline app/controllers/purchase_order_controller.rb]: space

File: app/controllers/purchase_order_controller.rb [hunk 7]
```diff
@@ -117,6 +115,10 @@ def confirm
 
   def deliver
     if @purchase_order.can_deliver?
+      @purchase_order.purchase_details.each do |pd|
```
Comment (sulemanuzair) [inline app/controllers/purchase_order_controller.rb]: use descriptive variable names

File: app/controllers/bills_controller.rb (hunk 3 above)
Comment (tariq-9484) [reply app/controllers/bills_controller.rb:3]: it isnt required anymore

File: app/controllers/purchase_order_controller.rb [hunk 8]
```diff
@@ -117,6 +115,10 @@ def confirm
...
+        medicine = pd.medicine
+        current_hospital.medicines.find_by(id: medicine.id).update(quantity: medicine.quantity+pd.quantity)
```
Comment (sulemanuzair) [inline app/controllers/purchase_order_controller.rb]: we have the same medicine object already, no need to find it from hospital again

Comment (sulemanuzair) [reply app/controllers/purchase_order_controller.rb]: space 
medicine.quantity + pd.quantity
pd should be purchase_detail

File: app/models/bill.rb [hunk 9]
```diff
@@ -8,24 +8,29 @@ class Bill < ApplicationRecord
...
-      if medicine.update(quantity: medicine.quantity-quantity_added)
-        update(price: price+=medicine.price*quantity_added)
-        curr_bill_detail=bill_details.find_by(billable: medicine)
-        if curr_bill_detail
-          curr_bill_detail.update(quantity:quantity_added+curr_bill_detail.quantity)
+    begin
+      Bill.transaction do
+        if medicine.quantity>0 && quantity_added<=medicine.quantity
+          if medicine.update(quantity: medicine.quantity-quantity_added)
+            update(price: self.price+=medicine.price*quantity_added)
```
Comment (sulemanuzair) [inline app/models/bill.rb]: calculate price separately and update through that calculation
e.g
updated_price = price + (medicine.price * quantity_added)
update(price: updated_price)

File: app/models/bill.rb [hunk 10]
```diff
@@ -8,24 +8,29 @@ class Bill < ApplicationRecord
...
-          bill_details.create(quantity: quantity_added, billable:medicine, hospital: medicine.hospital) 
-        end       
+          errors.add(:unable_to_add, I18n.t('medicine.add.failure'))
+          return false
+        end
       end
-    else
-      errors.add(:unable_to_add, I18n.t('medicine.add.failure'))
-      return false
+      rescue ActiveRecord::RecordNotSaved
```
Comment (sulemanuzair) [inline app/models/bill.rb]: may be add some error in current object if exception is catched
or return false

File: app/models/purchase_order.rb [hunk 11]
```diff
@@ -22,19 +22,21 @@ class PurchaseOrder < ApplicationRecord
...
-        end       
+          self.errors.add(:unable_to_add, I18n.t('medicine.add.failure'))
+          return false 
+        end 
       end
-    else
-      self.errors.add(:unable_to_add, I18n.t('medicine.add.failure'))
-      return false
-    end
+      rescue ActiveRecord::RecordNotSaved
```
Comment (sulemanuzair) [inline app/models/purchase_order.rb]: similar changes required

File: app/models/bill.rb [hunk 12]
```diff
@@ -8,24 +8,29 @@ class Bill < ApplicationRecord
...
-    else
-      errors.add(:unable_to_add, I18n.t('medicine.add.failure'))
-      return false
+      rescue ActiveRecord::RecordNotSaved 
     end
   end
 
   def add_doctor(doctor)
-    if update(price: price+=doctor.consultancy_fee)
+    if update(price: self.price+=doctor.consultancy_fee)
```
Comment (sulemanuzair) [inline app/models/bill.rb]: space, why are we explicitly using self here

File: app/models/purchase_order.rb [hunk 13]
```diff
@@ -22,19 +22,21 @@ class PurchaseOrder < ApplicationRecord
...
   def add_medicine(medicine, quantity_added)
-    if medicine.quantity>0 && quantity_added<=medicine.quantity
-      if medicine.update(quantity: medicine.quantity-quantity_added)
-        self.update(price: self.price+=medicine.price*quantity_added)
-        curr_purchase_detail=purchase_details.find_by(medicine: medicine)
-        if curr_purchase_detail
-          curr_purchase_detail.update(quantity:quantity_added+curr_purchase_detail.quantity)
+    begin
+      PurchaseOrder.transaction do
+        if self.update(price: self.price+=medicine.price*quantity_added)
```
Comment (sulemanuzair) [inline app/models/purchase_order.rb:27]: do calculation and updation separately for complex cases

File: app/views/bills/show.html.erb [hunk 14]
```diff
@@ -5,9 +5,8 @@
           <h3 class="col-md-8"><%= @bill.patient.name %></h3> 
           <div class="col-md-4 d-flex justify-content-md-end"> 
             <%= link_to t('back'), { controller: 'bills', action: 'index'}, class: "btn btn-secondary" %> 
-            <%= link_to edit_bill_path(@bill), class: "btn btn-primary btn-spacing" do %> 
-              <i class="fas fa-pen"></i>
-            <%end%>
+            <%= link_to t('delete'), bill_path(@bill), method: :delete, class: "btn btn-danger btn-spacing", data: {confirm:'Are you sure you want to delete medicine?'} %>
```
Comment (sulemanuzair) [inline app/views/bills/show.html.erb:8]: space in data field

File: app/views/medicines/index.html.erb [hunk 15]
```diff
@@ -9,41 +9,45 @@
...
+              <td><%= link_to medicine.name, medicine_path(medicine) %></td>   
+              <td><%= medicine.price %></td>   
+              <td><%= medicine.quantity%></td>
+              <div>   
+                <td>   
+                  <%= link_to edit_medicine_path(medicine) do%>
+                    <i class="fas fa-pen"></i>
+                  <%end%> 
+                  <% if can? :destroy, medicine %>
+                    <%= link_to medicine_path(medicine), method: :delete, class: "btn-danger btn-spacing", data: {confirm:'Are you sure you want to delete medicine?'} do%>
```
Comment (sulemanuzair) [inline app/views/medicines/index.html.erb]: space
data: {confirm:'Are you sure you want to delete medicine?'} -> data: { confirm: 'Are you sure you want to delete medicine?' }

File: app/views/medicines/show.html.erb [hunk 16]
```diff
@@ -5,9 +5,10 @@
           <h3 class="col-md-8"><%= @medicine.name %></h3> 
           <div class="col-md-4 d-flex justify-content-md-end"> 
             <%= link_to t('back'), { controller: 'medicines', action: 'index'}, class: "btn btn-secondary " %>
-            <%= link_to edit_medicine_path(@medicine), class: "btn btn-primary btn-spacing" do %>
-              <i class="fas fa-pen"></i>
-            <%end%> 
+            <% if can? :destroy, @medicine %>
+              <%= link_to t('delete'), medicine_path(@medicine), method: :delete, class: "btn btn-danger btn-spacing", data: {confirm:'Are you sure you want to delete medicine?'}%>
```
Comment (sulemanuzair) [inline app/views/medicines/show.html.erb]: space

File: app/views/purchase_order/index.html.erb [hunk 17]
```diff
@@ -9,41 +9,57 @@
...
+              <th><%= t('purchase_order.state_label') %></th> 
+              <th class='action'><%= t('action') %></th>
+            </tr>   
+          </thead>   
+          <tbody>   
+          <% @purchase_orders.each do |po| %>   
+            <tr>   
+              <td><%=link_to po.vendorname, purchase_order_path(po)%></td>   
+              <td><%= po.price%></td>
+              <% if po.state == 'drafted' %>
```
Comment (sulemanuzair) [inline app/views/purchase_order/index.html.erb]: make a helper function for it

File: config/environments/test.rb [hunk 18]
```diff
@@ -41,7 +41,7 @@
   # ActionMailer::Base.deliveries array.
 
   # Print deprecation notices to the stderr.
-  config.active_support.deprecation = :stderr
+  config.active_support.deprecation = :log
```
Comment (sulemanuzair) [inline config/environments/test.rb:44]: why are we changing this ?

File: config/initializers/app_constants.rb [hunk 19]
```diff
@@ -1,2 +1,2 @@
 # frozen_string_literal: true
-PAGINATION_SIZE = 10
\ No newline at end of file
+PAGINATION_SIZE = 5
```
Comment (sulemanuzair) [inline config/initializers/app_constants.rb]: revert

File: spec/models/bill_spec.rb [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+      @doctor=@hospital.doctors.create(@doctor_params)
+      @bill_params = FactoryGirl.attributes_for(:bills)
+      @bill_params[:hospital] = @hospital
+      @bill_params[:patient] = @patient
+      @bill = @hospital.bills.create(@bill_params)
+    end
+
+    it 'should be able to add medicines to itself' do
+      @bill.billable_type = 'medicine'
+      @medicine.quantity = 10
```
Comment (sulemanuzair) [inline spec/models/bill_spec.rb:44]: there should not be calculations in 'it' block
suggestion: make a describe method and change values in its before action, then test in 'it' block

File: config/environments/test.rb (hunk 18 above)
Comment (tariq-9484) [reply config/environments/test.rb:44]: to log test case warnings to the log file instead of console

File: spec/models/bill_spec.rb (hunk 20 above)
Comment (tariq-9484) [reply spec/models/bill_spec.rb:44]: These were required for the test case to run successfully. i did it after discussing with sir hamza

**Pull request merged**